### PR TITLE
SWATCH-1230: Implement swatch-product-configuration lookup methods

### PR DESCRIPTION
**Pull request by lindseyburnett**
Jira issue: [SWATCH-1230](https://issues.redhat.com/browse/SWATCH-1230)

Very important note: The term "Subscription" in the context of this PR (and swatch-product-configuration module in general) is NOT the same as the term "subscription" in swatch-contracts or swatch-subscription-sync.  The terms in this library mirror terms defined by the product management team.  Please see the doc attached to the epic for this card for more information.  There's also short explanations at the top of the java classes.

## Highlights
Created a List<Subscription> that is populated from the yaml config files, provide some common lookup methods, and ensure that Subscriptions can be looked up from other swatch modules.  We only want to read from the disk to populate this list if we haven't done so already.

Moved "POJO" classes from `com.redhat.swatch.configuration.model` to `com.redhat.swatch.configuration.registry`.  Why?  Added a `SubscriptionConfigRegistry` class that I wanted to be be package private, and accessible from the "POJO" classes.  Also, the Subscription/Variant started to have lookup methods that made the POJOs not really Plain Old Java Objects anymore. (Actually done in PR #2307 )

Introduced the `src/main/resource/subscription_configs` directory.  Why?  I liked the idea of having to check/maintain the classpath for a single `subscription_configs` string for yaml files, instead of maintaining a `List.of("OpenShift", "RHEL")` in the code.

YAML properties changed to camelCase instead of snake_case due to a limitation with Snake YAML library.

## Testing
There isn't a great way to test these lookup methods outside of unit tests.  Below I'll provide some steps to prove out how these lookup methods can be accessed from the other swatch modules.

### Save & apply this git patch
0e9e34.patch
```
diff --git a/swatch-producer-aws/src/main/java/com/redhat/swatch/resource/BillableUsageResource.java b/swatch-producer-aws/src/main/java/com/redhat/swatch/resource/BillableUsageResource.java
index 6ac854b..9f4ae0c 100644
--- a/swatch-producer-aws/src/main/java/com/redhat/swatch/resource/BillableUsageResource.java
+++ b/swatch-producer-aws/src/main/java/com/redhat/swatch/resource/BillableUsageResource.java
@@ -20,10 +20,12 @@
  */
 package com.redhat.swatch.resource;
 
+import com.redhat.swatch.configuration.registry.Subscription;
 import com.redhat.swatch.exception.AwsManualSubmissionDisabledException;
 import com.redhat.swatch.openapi.model.BillableUsage;
 import com.redhat.swatch.openapi.resource.DefaultApi;
 import com.redhat.swatch.processors.BillableUsageProducer;
+import lombok.SneakyThrows;
 import lombok.extern.slf4j.Slf4j;
 import org.eclipse.microprofile.config.inject.ConfigProperty;
 
@@ -40,12 +42,16 @@ public class BillableUsageResource implements DefaultApi {
   }
 
   @Override
+  @SneakyThrows
   public void submitBillableUsage(BillableUsage billableUsage) {
     if (!manualSubmissionEnabled) {
       throw new AwsManualSubmissionDisabledException();
     }
+
     log.info("{}", billableUsage);
 
+    log.info("{}", Subscription.findById("basilisk").get());
+
     tallySummaryProducer.queueBillableUsage(billableUsage);
   }
 }
diff --git a/swatch-producer-aws/src/main/resources/application.properties b/swatch-producer-aws/src/main/resources/application.properties
index ab28001..bcd6df7 100644
--- a/swatch-producer-aws/src/main/resources/application.properties
+++ b/swatch-producer-aws/src/main/resources/application.properties
@@ -7,7 +7,7 @@ AWS_MANUAL_SUBMISSION_ENABLED=false
 AWS_SEND_RETRIES=0
 AWS_USAGE_CONTEXT_LOOKUP_RETRIES=0
 AWS_MARKETPLACE_ENDPOINT_OVERRIDE=false
-ENABLE_SPLUNK_HEC=true
+ENABLE_SPLUNK_HEC=false
 SPLUNK_HEC_URL=https://splunk-hec.redhat.com:8088/
 SPLUNK_SOURCE=swatch-producer-aws
 SPLUNK_SOURCE_TYPE=quarkus_service
```

### Start swatch-producer-aws 
`DEV_MODE=true ./gradlew :swatch-producer-aws:quarkusDev`

### Hit this curl command
```bash
curl -X 'POST' \
  'http://localhost:8000/api/swatch-producer-aws/internal/aws/billable_usage' \
  -H 'accept: */*' \
  -H 'Content-Type: application/json' \
  -d '{
  "account_number": "string",
  "org_id": "string",
  "id": "3fa85f64-5717-4562-b3fc-2c963f66afa6",
  "billing_provider": "red hat",
  "billing_account_id": "string",
  "snapshot_date": "2022-03-10T12:15:50-04:00",
  "product_id": "string",
  "sla": "Premium",
  "usage": "Production",
  "uom": "Sockets",
  "value": 0,
  "billing_factor": 0,
  "vendor_product_code": "string"
}'
```

### Note output in log

Side note: You should only see the message "Attempting to walk directoryPath:" the first time you hit the curl command, since we don't want to repeatedly load from the disk every time.

`2023-07-11 12:49:22,934 DEBUG [com.red.swa.con.reg.SubscriptionConfigRegistry] (executor-thread-0) Attempting to walk directoryPath: /home/lburnett/code/rhsm-subscriptions/swatch-product-configuration/build/resources/main/subscription_configs`
`2023-07-11 12:49:22,965 INFO  [com.red.swa.res.BillableUsageResource] (executor-thread-0) Subscription(platform=OpenShift, id=basilisk, parentSubscription=null, includedSubscriptions=[], fingerprint=null, variants=[Variant(tag=basilisk, roles=[BASILISK], engineeringIds=[], productNames=[BASILISK])], billingWindow=MONTHLY, serviceType=BASILISK Instance, metrics=[Metric(id=redhat.com:BASILISK:transfer_gb, rhmMetricId=redhat.com:BASILISK:transfer_gb, awsDimension=transfer_gb, prometheus=PrometheusMetric(queryKey=null, queryParams={product=BASILISK, metric=kafka_id:haproxy_server_bytes_in_out_total:rate1h_gibibytes, metadata_metric=subscription_labels}), billingFactor=null), Metric(id=redhat.com:BASILISK:cluster_hour, rhmMetricId=redhat.com:BASILISK:cluster_hour, awsDimension=cluster_hour, prometheus=PrometheusMetric(queryKey=null, queryParams={product=BASILISK, metric=kafka_id:strimzi_resource_state:max_over_time1h, metadata_metric=subscription_labels}), billingFactor=null), Metric(id=redhat.com:BASILISK:storage_gib_months, rhmMetricId=redhat.com:BASILISK:storage_gib_months, awsDimension=storage_gb, prometheus=PrometheusMetric(queryKey=null, queryParams={product=BASILISK, metric=kafka_id:kafka_broker_quota_totalstorageusedbytes:max_over_time1h_gibibyte_months, metadata_metric=subscription_labels}), billingFactor=null)], defaults=Defaults(variant=basilisk, sla=PREMIUM, usage=PRODUCTION))`

### CODE REVIEW

File: swatch-product-configuration/build.gradle [hunk 1]
```diff
@@ -4,11 +4,12 @@ plugins {
 }
 
 dependencies {
-    compileOnly platform(libraries["quarkus-bom"])
```
Comment (lindseyburnett) [inline swatch-product-configuration/build.gradle:7]: Including quarkus-bom was for snakeyaml and slf4j.  Enforcing the quarkus-bom library was causing issues when trying to run the spring boot modules.  Opted to remove references to quarkus-bom and explicitly set the version number.

File: swatch-product-configuration/src/main/java/com/redhat/swatch/configuration/registry/Metric.java [hunk 2]
```diff
@@ -18,19 +18,18 @@
...
 
 @Data
-@Builder
 public class Metric {
 
   @NotNull @NotEmpty private String id; // required
   private String rhmMetricId;
   private String awsDimension;
-  private PrometheusMetric prometheusMetric;
+  private PrometheusMetric prometheus;
```
Comment (lindseyburnett) [inline swatch-product-configuration/src/main/java/com/redhat/swatch/configuration/registry/Metric.java:33]: Renamed to match what we named this in the yaml files

**Comment by kartikshahc** [comment]
@lindseyburnett Did we decide to use `com.resources.my_package`(Snake Case using underscore) for package? And camel casing in yaml properties?

File: swatch-product-configuration/src/main/java/com/redhat/swatch/configuration/registry/Subscription.java [hunk 3]
```diff
@@ -43,21 +45,69 @@ public class Subscription {
...
+
+    return this.getMetrics().stream().anyMatch(metric -> Objects.nonNull(metric.getPrometheus()));
+  }
+
+  public String getFinestGranularity() {
+
+    return this.isPrometheusEnabled() ? "HOURLY" : "DAILY";
+  }
+
+  public List<String> getSupportedGranularity() {
```
Comment (kartikshahc) [inline swatch-product-configuration/src/main/java/com/redhat/swatch/configuration/registry/Subscription.java:108]: Could we also change in the Jira card the method name `getSupportedGranularity` instead of `supportsGranularity`? I don't mind if we want to keep it the way it is currently in Jira.

File: swatch-product-configuration/src/main/java/com/redhat/swatch/configuration/registry/Variant.java [hunk 4]
```diff
@@ -32,11 +34,25 @@
...
+  }
+
+  public static Optional<Variant> findByRole(String role) {
+    return SubscriptionConfigRegistry.getInstance()
+        .lookupSubscriptionByRole(role)
+        .flatMap(
+            subscription ->
+                subscription.getVariants().stream()
+                    .filter(variant -> !variant.getRoles().isEmpty())
+                    .filter(variant -> variant.getRoles().contains(role))
```
Comment (kartikshahc) [inline swatch-product-configuration/src/main/java/com/redhat/swatch/configuration/registry/Variant.java:55]: It will be better if you can combine two filters in one like below:
```
SubscriptionConfigRegistry.getInstance()
    .lookupSubscriptionByRole(role)
    .flatMap(subscription -> subscription.getVariants().stream()
        .filter(variant -> !variant.getRoles().isEmpty() && variant.getRoles().contains(role))
        .findFirst());
```

File: swatch-product-configuration/src/main/java/com/redhat/swatch/configuration/registry/SubscriptionConfigRegistry.java [hunk 5]
```diff
@@ -0,0 +1,242 @@
...
+
+  /**
+   * @param serviceType
+   * @return Optional<Subscription>
+   */
+  Optional<Subscription> lookupByServiceType(String serviceType) {
+
+    return subscriptions.stream()
+        .filter(subscription -> Objects.nonNull(subscription.getServiceType()))
+        .filter(subscription -> Objects.equals(subscription.getServiceType(), serviceType))
```
Comment (kartikshahc) [inline swatch-product-configuration/src/main/java/com/redhat/swatch/configuration/registry/SubscriptionConfigRegistry.java:129]: I've removed the first filter() condition Objects.nonNull(subscription.getServiceType()) because Objects.equals() already handles null-checking. So, there's no need to explicitly filter out null values for the service type. 

```
subscriptions.stream()
    .filter(subscription -> Objects.equals(subscription.getServiceType(), serviceType))
    .findFirst();
```

File: swatch-product-configuration/src/main/java/com/redhat/swatch/configuration/registry/Subscription.java [hunk 6]
```diff
@@ -43,21 +45,69 @@ public class Subscription {
...
+        .findFirst();
+  }
+
+  public static List<String> getAllServiceTypes() {
+    return SubscriptionConfigRegistry.getInstance().getAllServiceTypes();
+  }
+
+  public boolean isPrometheusEnabled() {
+
+    return this.getMetrics().stream().anyMatch(metric -> Objects.nonNull(metric.getPrometheus()));
```
Comment (kartikshahc) [inline swatch-product-configuration/src/main/java/com/redhat/swatch/configuration/registry/Subscription.java:100]: Since you are checking at some places if metrics is non null do should we check here by doing `.filter(Objects::nonNull)`? 
Although I don't see a reason doing that because each of the config file contains metric. I will leave upto you on whether to remove from above or add the check here.

**Review by kartikshahc — CHANGES_REQUESTED** [review]

**Comment by kartikshahc** [comment]
Tested the Subscription.findById test mentioned in PR and works fine. I also tested all the methods with negative test cases and works fine.
```
Subscription(platform=OpenShift, id=rosa, parentSubscription=null, includedSubscriptions=[], fingerprint=null, variants=[Variant(tag=rosa, roles=[moa-hostedcontrolplane], engineeringIds=[], productNames=[OpenShift Dedicated])], billingWindow=MONTHLY, serviceType=rosa Instance, metrics=[Metric(id=redhat.com:rosa:cluster_hour, rhmMetricId=null, awsDimension=control_plane, prometheus=PrometheusMetric(queryKey=null, queryParams={product=moa-hostedcontrolplane, metric=cluster:usage:workload:capacity_physical_instance_hours, metadata_metric=subscription_labels}), billingFactor=null), Metric(id=redhat.com:rosa:cpu_hour, rhmMetricId=null, awsDimension=four_vcpu_hour, prometheus=PrometheusMetric(queryKey=null, queryParams={product=moa-hostedcontrolplane, metric=cluster:usage:workload:capacity_physical_cpu_hours, metadata_metric=subscription_labels}), billingFactor=0.25)], defaults=Defaults(variant=rosa, sla=PREMIUM, usage=PRODUCTION))


Subscription(platform=OpenShift, id=OpenShift-dedicated-metrics, parentSubscription=null, includedSubscriptions=[], fingerprint=null, variants=[Variant(tag=OpenShift-dedicated-metrics, roles=[osd], engineeringIds=[], productNames=[OpenShift Dedicated])], billingWindow=MONTHLY, serviceType=OpenShift Cluster, metrics=[Metric(id=redhat.com:openshift_dedicated:4cpu_hour, rhmMetricId=redhat.com:openshift_dedicated:4cpu_hour, awsDimension=null, prometheus=PrometheusMetric(queryKey=default, queryParams={product=osd, metric=cluster:usage:workload:capacity_physical_cpu_hours, metadata_metric=subscription_labels}), billingFactor=0.25), Metric(id=redhat.com:openshift_dedicated:cluster_hour, rhmMetricId=redhat.com:openshift_dedicated:cluster_hour, awsDimension=null, prometheus=PrometheusMetric(queryKey=default, queryParams={product=osd, metric=cluster:usage:workload:capacity_physical_instance_hours, metadata_metric=subscription_labels}), billingFactor=null)], defaults=Defaults(variant=OpenShift-dedicated-metrics, sla=PREMIUM, usage=PRODUCTION))

2023-07-14 07:35:33,116 INFO  [com.red.swa.res.BillableUsageResource] (executor-thread-0) Subscription(platform=OpenShift, id=OpenShift Container Platform, parentSubscription=null, includedSubscriptions=[], fingerprint=Fingerprint(engineeringIds=[290], arches=[]), variants=[Variant(tag=OpenShift Container Platform, roles=[], engineeringIds=[], productNames=[])], billingWindow=null, serviceType=null, metrics=[Metric(id=SOCKETS, rhmMetricId=null, awsDimension=null, prometheus=null, billingFactor=null), Metric(id=CORES, rhmMetricId=null, awsDimension=null, prometheus=null, billingFactor=null)], defaults=Defaults(variant=OpenShift Container Platform, sla=null, usage=null))

2023-07-14 07:35:33,116 INFO  [com.red.swa.res.BillableUsageResource] (executor-thread-0) Subscription(platform=OpenShift, id=OpenShift-metrics, parentSubscription=null, includedSubscriptions=[], fingerprint=null, variants=[Variant(tag=OpenShift-metrics, roles=[ocp], engineeringIds=[], productNames=[OpenShift Container Platform])], billingWindow=MONTHLY, serviceType=OpenShift Cluster, metrics=[Metric(id=redhat.com:openshift_container_platform:cpu_hour, rhmMetricId=redhat.com:openshift_container_platform:cpu_hour, awsDimension=null, prometheus=PrometheusMetric(queryKey=default, queryParams={product=ocp, metric=cluster:usage:workload:capacity_physical_cpu_hours, metadata_metric=subscription_labels}), billingFactor=null)], defaults=Defaults(variant=OpenShift-metrics, sla=PREMIUM, usage=PRODUCTION))

2023-07-14 07:45:47,690 INFO  [com.red.swa.res.BillableUsageResource] (executor-thread-0) Subscription(platform=RHEL, id=RHEL for x86, parentSubscription=null, includedSubscriptions=[], fingerprint=Fingerprint(engineeringIds=[479], arches=[i386, x86_64]), variants=[Variant(tag=RHEL Desktop, roles=[RHEL Desktop], engineeringIds=[68], productNames=[]), Variant(tag=RHEL Server, roles=[Red Hat Enterprise Linux Server], engineeringIds=[69], productNames=[]), Variant(tag=RHEL Workstation, roles=[Red Hat Enterprise Linux Workstation], engineeringIds=[71], productNames=[]), Variant(tag=RHEL Compute Node, roles=[Red Hat Enterprise Linux Compute Node], engineeringIds=[76], productNames=[])], billingWindow=null, serviceType=null, metrics=[Metric(id=SOCKETS, rhmMetricId=null, awsDimension=null, prometheus=null, billingFactor=null)], defaults=Defaults(variant=RHEL Server, sla=null, usage=null))
```

File: CONTRIBUTING.md [hunk 7]
```diff
@@ -41,3 +41,13 @@ There are various other tasks, see `./gradlew help tasks | grep liquibase`. Note
...
 
 DB migrations should be written to roll back cleanly (exceptions should be discussed with the team).
+
+-------------
+
+# swatch-product-configuration
+
+Directory structure: `$Platform/$Subscription_Name.yaml` (e.g. RHEL/RHEL_for_x86.yaml).
+
+Use camel case for yaml properties that need to be serialized into the `com.redhat.swatch.configuration.registry` object definitions.
```
[the remark this answers is not in the document]
Comment (lindseyburnett) [reply CONTRIBUTING.md:51]: File naming convention came from the epic design/rationale doc.  All the Tags were listed with the expected tags and what file they should be in.

> note: underscores instead of spaces as a convention

https://docs.google.com/document/d/1qrYBxqbf9FOBKWAOYD6RDyQoj8AzNomF4cHRkAEhc9o/edit#heading=h.c52vkgk01ksn

File: swatch-product-configuration/src/main/java/com/redhat/swatch/configuration/registry/Subscription.java (hunk 6 above)
Comment (lindseyburnett) [reply swatch-product-configuration/src/main/java/com/redhat/swatch/configuration/registry/Subscription.java:100]: To be honest, I have no idea why I have that `.filter(Objects::nonNull)` on line 77.  I can't think of a time where `this.getMetrics()` would ever contain a null - even if there's no metrics in a config file, `metrics` would initialize as an empty list.

I'm going to remove the unnecessary `.filter(Objects::nonNull)` check

File: swatch-product-configuration/src/main/resources/subscription_configs/OpenShift/rhosak.yaml [hunk 8]
```diff
@@ -20,31 +20,31 @@ serviceType: Kafka Cluster
...
   - id: redhat.com:rhosak:storage_gib_months
-    rhm_metric_id: redhat.com:rhosak:storage_gib_months
-    aws_dimension: storage_gb
+    rhmMetricId: redhat.com:rhosak:storage_gib_months
+    awsDimension: storage_gb
     prometheus:
       queryParams:
         product: rhosak
         metric: kafka_id:kafka_broker_quota_totalstorageusedbytes:max_over_time1h_gibibyte_months
         metadata_metric: subscription_labels
```
[the remark this answers is not in the document]
Comment (lindseyburnett) [reply swatch-product-configuration/src/main/resources/subscription_configs/OpenShift/rhosak.yaml:45]: I think that one needs to stay the way it is because it's part of the query params sent to the prometheus api?

https://github.com/RedHatInsights/rhsm-subscriptions/pull/2306/files#diff-eca12c0a30e25b4b46522ebf89465a03ba72a03f540796c979137931d8f92055R53

File: swatch-product-configuration/src/main/java/com/redhat/swatch/configuration/registry/Subscription.java (hunk 3 above)
Comment (lindseyburnett) [reply swatch-product-configuration/src/main/java/com/redhat/swatch/configuration/registry/Subscription.java:108]: Done - renamed in the JIRA card description

File: swatch-product-configuration/src/main/java/com/redhat/swatch/configuration/registry/Variant.java [hunk 9]
```diff
@@ -32,11 +34,25 @@
...
-  private List<String> engineeringIds;
-  private List<String> productNames;
+  private List<String> roles = new ArrayList<>();
+  private List<String> engineeringIds = new ArrayList<>();
+  private List<String> productNames = new ArrayList<>();
+
+  public Optional<Subscription> getSubscription() {
+    return SubscriptionConfigRegistry.getInstance().lookupSubscriptionByTag(tag);
+  }
+
```
[the remark this answers is not in the document]
Comment (lindseyburnett) [reply swatch-product-configuration/src/main/java/com/redhat/swatch/configuration/registry/Variant.java:47]: The important part of this "lookup methods" cards was less about getting every necessary lookup method implemented, and more about getting the values read from disk and putting the pattern of lookup methods in place.

Had a discussion with @kahowell last week and decided that I was going to put the PR up for the ones I already defined.  As the other cards in the epic get worked on, lookup methods can be added as needed.  Coincidently, I'm  implementing the findByEngProductId in the next card because I ended up needing it.

I'll remove the two methods you mentioned from the JIRA card description.

File: swatch-product-configuration/src/main/resources/subscription_configs/OpenShift/rhosak.yaml (hunk 8 above)
[the remark this answers is not in the document]
Comment (lindseyburnett) [reply swatch-product-configuration/src/main/resources/subscription_configs/OpenShift/rhosak.yaml]: Update - these don't directly get sent to the prometheus api.  They're used to inject values into the query template coming from the `src/main/resources/application-openshift-metering-worker.yaml`.

Updated the [SWATCH-1234 Refactor prometheus integration to use swatch-product-configuration](https://issues.redhat.com/browse/SWATCH-1234) card to call this out

File: swatch-product-configuration/src/main/java/com/redhat/swatch/configuration/registry/SubscriptionConfigRegistry.java [hunk 10]
```diff
@@ -0,0 +1,242 @@
...
+ * lookup methods for that list.
+ */
+@Slf4j
+class SubscriptionConfigRegistry {
+
+  private static SubscriptionConfigRegistry instance = null;
+
+  @Getter private final List<Subscription> subscriptions;
+
+  public static synchronized SubscriptionConfigRegistry getInstance() {
```
[the remark this answers is not in the document]
Comment (lindseyburnett) [reply swatch-product-configuration/src/main/java/com/redhat/swatch/configuration/registry/SubscriptionConfigRegistry.java]: Timeboxed the Bean/Singleton/DI endeavor and it ended up being too complicated to implement now.

Updated SubscriptionConfigRegistry -> SubscriptionRegistry and removed the passthrough methods

**Review by kartikshahc — APPROVED** [review]
